src/qtviewer/ParseIfcFile.cpp

```diff
@@ -21,22 +31,145 @@ void ParseIfcFile::Parse(const std::string& filePath)
 
     IfcGeom::Iterator* it = new IfcGeom::Iterator(settings, &file);
     if (!it->initialize()) {
-        outputMsg("Error: Iterator failed to initialize! Aborting.");
+        MessageLogger::log("Error: Iterator failed to initialize! Aborting.");
         delete it;
-        return;
+        return false;
     }
 
     do {
         //const IfcGeom::BRepElement* bRepElem = it->get_native();
         const IfcGeom::TriangulationElement* triElem = static_cast<const IfcGeom::TriangulationElement*>(it->get());
```

Comment:
I don't see you use. `triElem->transformation().matrix().data()`. This is the 4x3 matrix that would position the element in world coordinates. You can use this to populate an Matrixd and MatrixTransform.

What's more: if you look at triElem->geometry().id() you might see that the id is equal to the previous iteration. In that case you don't need to create a new osg::Geometry at all, but reuse the last one, but with a a different ElementTransform parent.

Reply:
I tried it for different IFCs and found that the matrix data was the same for all the elements:
```
1.000000  0.000000  0.000000  0.000000  
1.000000  0.000000  0.000000  0.000000  
1.000000  0.000000  0.000000  0.000000 
```
Not sure what I'm missing here.

The sequence of triElem->geometry().id() values is like this:
```
355-world-coords
14488-world-coords
15016-world-coords
17018-openings-17111-world-coords
17095-world-coords
17456-world-coords
18443-world-coords
18676-openings-18771-18839-world-coords
18757-world-coords
18825-world-coords
19187-world-coords
19492-world-coords
20360-world-coords
20576-world-coords
...
```
So they don't seem to be consecutive. Please explain what you meant to suggest, based on these outputs. Thanks.

Reply:
Yep, sorry, see https://github.com/IfcOpenShell/IfcOpenShell/pull/3739/files#r1355661970

Also it helps to print the matrix as

```
1.000000  0.000000  0.000000  
0.000000  1.000000  0.000000  
0.000000  0.000000  1.000000
0.000000  0.000000  0.000000
```` 

3 columns instead of 4. When passing the data to OpenGL you probably need to add a fourth column

```
0.000000
0.000000
0.000000
1.000000
```

So that you get a 4x4 square matrix.

(This 3x4 matrix is what opencascade does, so we copied it).

Reply:
Hi,
Setting USE_WORLD_COORDS to false does give the coordinates in the transformation matrix data. The triElem->geometry().id()s are still not consecutive. They seem to be in ascending order. For some IFCs they are just numbers while for some they are a string like:
```
32-material-89
34-material-89
35-material-99
37-material-89
38-material-99
```
In the current method, new osg::Geometry is required to create new geometry nodes for OSG. Please suggest how you were proposing to do it in a single osg::Geometry node. And would that affect the selection of elements in the model if they are all a single node?

Reply:
You can try with Duplex_A https://portal.nibs.org/files/wl/?id=O9inoVWDytV4P3UTS7iIdqhVKcTyVlcs

What may also help is to convert the ifc to .dae and import it in Blender (just plain Blender, no need for BlenderBIM here) to have a look. The collada (.dae) serializer operates just on the same data.

**Indeed as you observed quite a few elements with just a numeric representation name**

![afbeelding](https://github.com/IfcOpenShell/IfcOpenShell/assets/1096535/a89bafec-a47a-4b4b-86c5-32c87533f75b)

**Others with openings or materials applied get that postfixed**

![afbeelding](https://github.com/IfcOpenShell/IfcOpenShell/assets/1096535/cfa00d5c-b02d-490e-826a-2b3f35237da0)

**But again others, typically things like windows and doors have that representation shared. Note the [`4`] next to the "representation-5994" box. When you edit that representation in Blender you can see that you're editing 4 element representations at once.**

![afbeelding](https://github.com/IfcOpenShell/IfcOpenShell/assets/1096535/1b021352-5289-4ae6-97c6-b5f30a94059d)

---

Regarding OSG I'm not an expert, but I think you'd create multiple osg::MatrixTransform that basically represent the Object (or rather it's placement) and then assign osg::Geometry to these. In case of reused shapes you'd assign the same osg::Geometry to multiple osg::MatrixTransform.

Reply:
Thanks for the reference model and the images. Btw how did you convert the IFC to Collada? I used Archicad to convert it to .dae and then imported it in Blender. All the nodes are numbered serially for me in a consistent format:
<img width="1570" alt="Screenshot 2023-12-19 at 7 02 54 PM" src="https://github.com/IfcOpenShell/IfcOpenShell/assets/39452934/21594d6b-e396-4873-8a29-f7b15e38a411">

Reply:
Updated code. _(Using transformation-matrix data; Turned off world-coords; Re-using geometry of repeating IDs)_